### DISCUSSION TITLE
Termination Criterion

### DISCUSSION THREAD

**Discussion by m-zakeri** [anyoptimization/pymoo#235] [Q&A] [answered]
How can I understand with which condition(s) is my algorithm terminated when I use `MultiObjectiveDefaultTermination` (with customized values for arguments )?
In other words, I want to print the first criterion which causes the termination of my algorithm.
I can print the number of executed generations by `res.algorithm.n_gen`, but what about other criteria such as `f_tol`, `cv_tol`, etc.

**Comment by blankjul** [comment, marked as answer]
So far, the termination does not store why the decision has been made. The easiest way of achieving this is by inheriting from `MultiObjectiveDefaultTermination` and storing the decision in the object itself.
See the example below. Also, note that `cv_tol` is violated anytime for unconstrained problems to ensure the expression becomes false when `f_tol` is violated.

```
from pymoo.algorithms.moo.nsga2 import NSGA2
from pymoo.factory import get_problem
from pymoo.optimize import minimize
from pymoo.util.termination.default import MultiObjectiveDefaultTermination

problem = get_problem("zdt1")

algorithm = NSGA2(pop_size=100, eliminate_duplicates=True)

termination = MultiObjectiveDefaultTermination()


class MyMultiObjectiveDefaultTermination(MultiObjectiveDefaultTermination):

    def __init__(self, **kwargs) -> None:
        super().__init__(**kwargs)
        self.reasons = None

    def _decide(self, metrics):
        decisions = metrics[-1]
        cont = decisions["x_tol"] and (decisions["cv_tol"] or decisions["f_tol"])

        if not cont:
            self.reasons = [name for name in ["x_tol", "cv_tol", "f_tol"] if not decisions[name]]

        return cont


ret = minimize(problem,
               algorithm,
               termination=MyMultiObjectiveDefaultTermination(),
               seed=1,
               save_history=False,
               verbose=True)


print(ret.algorithm.termination.reasons)
```
